openml/runs/run.py

```diff
@@ -222,50 +249,84 @@ def get_metric_fn(self, sklearn_fn, kwargs={}):
             # TODO: make this a stream reader
         else:
-            raise ValueError('Run should have been locally executed or contain outputfile reference.')
+            raise ValueError('Run should have been locally executed or '
+                             'contain outputfile reference.')
+
+        # Need to know more about the task to compute scores correctly
+        task = get_task(self.task_id)
 
         attribute_names = [att[0] for att in predictions_arff['attributes']]
-        if 'correct' not in attribute_names:
-            raise ValueError('Attribute "correct" should be set')
-        if 'prediction' not in attribute_names:
-            raise ValueError('Attribute "predict" should be set')
+        if task.task_type == 'Supervised Classification' and 'correct' not in \
+                attribute_names:
+            raise ValueError('Attribute "correct" should be set for '
+                             'classification task runs')
+        if task.task_type == 'Supervised Regression' and 'truth' not in \
+                attribute_names:
+            raise ValueError('Attribute "truth" should be set for '
+                             'regression task runs')
+        if task.task_type != 'Clustering' and 'prediction' not in \
+                attribute_names:
+            raise ValueError('Attribute "predict" should be set for '
+                             'supervised task runs')
 
         def _attribute_list_to_dict(attribute_list):
-            # convenience function: Creates a mapping to map from the name of attributes
-            # present in the arff prediction file to their index. This is necessary
-            # because the number of classes can be different for different tasks.
+            # convenience function: Creates a mapping to map from the name of
+            # attributes present in the arff prediction file to their index.
+            # This is necessary because the number of classes can be different
+            # for different tasks.
             res = OrderedDict()
             for idx in range(len(attribute_list)):
                 res[attribute_list[idx][0]] = idx
             return res
-        attribute_dict = _attribute_list_to_dict(predictions_arff['attributes'])
 
-        # might throw KeyError!
-        predicted_idx = attribute_dict['prediction']
-        correct_idx = attribute_dict['correct']
+        attribute_dict = \
+            _attribute_list_to_dict(predictions_arff['attributes'])
+
         repeat_idx = attribute_dict['repeat']
         fold_idx = attribute_dict['fold']
-        sample_idx = attribute_dict['sample'] # TODO: this one might be zero
-
-        if predictions_arff['attributes'][predicted_idx][1] != predictions_arff['attributes'][correct_idx][1]:
+        predicted_idx = attribute_dict['prediction']  # Assume supervised tasks
+
+        if task.task_type == 'Supervised Classification' or \
```

Comment:
task_type_id

Reply:
Not sure what you mean. Are you referring to the same refactoring to enums as before?

Reply:
He means `if task.task_type_d == 1`. But I think it would be great to have `if task.task_type == constants.SUPERVISED_CLASSIFICATION`.